a few fixes and suggestions concerning groups

ThomasBreuer opened the pull request:
- fixed two errors in the documentation

- fixed `cokernel` (needs `normal_closure`)

- fixed the `show` method for conjugacy classes

- inserted an `IsIdenticalObj` check in `_maxgroup`
  (perhaps this is rather a GAP problem, but for the moment,
  the change seems to make sense)

- added `normal_closure` for two groups,
  under the assumption that the "small" group is a subgroup of the "big" one;
  GAP does *not* assume this, thus we have a problem here:
  The assumption should be checked, but this can be expensive;
  thus a user should at least have a chance to omit the check.
  And in order to get the flexibility of GAP's `NormalClosure`,
  something such as GAP's `ClosureGroup` can be introduced
  --I think we will need this anyhow, sooner or later.

- call `GAP.Globals.IsPrimeInt` instead of `GAP.Globals.IsPrime`,
  this avoids some overhead on the GAP side;
  since we know that the number in question is an `Int64`,
  a perhaps better alternative would be to call Nemo's `isprime`.

ThomasBreuer committed:
a few fixes and suggestions concerning groups

- fixed two errors in the documentation

- fixed `cokernel` (needs `normal_closure`)

- fixed the `show` method for conjugacy classes

- inserted an `IsIdenticalObj` check in `_maxgroup`
  (perhaps this is rather a GAP problem, but for the moment,
  the change seems to make sense)

- added `normal_closure` for two groups,
  under the assumption that the "small" group is a subgroup of the "big" one;
  GAP does *not* assume this, thus we have a problem here:
  The assumption should be checked, but this can be expensive;
  thus a user should at least have a chance to omit the check.
  And in order to get the flexibility of GAP's `NormalClosure`,
  something such as GAP's `ClosureGroup` can be introduced
  --I think we will need this anyhow, sooner or later.

- call `GAP.Globals.IsPrimeInt` instead of `GAP.Globals.IsPrime`,
  this avoids some overhead on the GAP side;
  since we know that the number in question is an `Int64`,
  a perhaps better alternative would be to call Nemo's `isprime`.

ThomasBreuer committed:
a few fixes and suggestions concerning groups

- fixed two errors in the documentation

- fixed `cokernel` (needs `normal_closure`)

- fixed the `show` method for conjugacy classes

- inserted an `IsIdenticalObj` check in `_maxgroup`
  (perhaps this is rather a GAP problem, but for the moment,
  the change seems to make sense)

- added `normal_closure` for two groups,
  under the assumption that the "small" group is a subgroup of the "big" one;
  GAP does *not* assume this, thus we have a problem here:
  The assumption should be checked, but this can be expensive;
  thus a user should at least have a chance to omit the check.
  And in order to get the flexibility of GAP's `NormalClosure`,
  something such as GAP's `ClosureGroup` can be introduced
  --I think we will need this anyhow, sooner or later.

- call `GAP.Globals.IsPrimeInt` instead of `GAP.Globals.IsPrime`,
  this avoids some overhead on the GAP side;
  since we know that the number in question is an `Int64`,
  a perhaps better alternative would be to call Nemo's `isprime`.

ThomasBreuer commented:
The two test jobs with Julia 1.3 fail with the following error message.
```
ERROR: LoadError: InitError: ccall: could not find function GAP_EvalString
Stacktrace:
 [1] evalstr_ex at /Users/travis/.julia/packages/GAP/tkivN/src/ccalls.jl:36 [inlined]
 [2] evalstr(::String) at /Users/travis/.julia/packages/GAP/tkivN/src/ccalls.jl:63
 [3] reset_GAP_ERROR_OUTPUT() at /Users/travis/.julia/packages/GAP/tkivN/src/GAP.jl:142
 [4] initialize(::Array{String,1}) at /Users/travis/.julia/packages/GAP/tkivN/src/GAP.jl:289
```
The other tests pass. What is going on here?

GDeFranceschi commented:
> The two test jobs with Julia 1.3 fail with the following error message.
> 
> ```
> ERROR: LoadError: InitError: ccall: could not find function GAP_EvalString
> Stacktrace:
>  [1] evalstr_ex at /Users/travis/.julia/packages/GAP/tkivN/src/ccalls.jl:36 [inlined]
>  [2] evalstr(::String) at /Users/travis/.julia/packages/GAP/tkivN/src/ccalls.jl:63
>  [3] reset_GAP_ERROR_OUTPUT() at /Users/travis/.julia/packages/GAP/tkivN/src/GAP.jl:142
>  [4] initialize(::Array{String,1}) at /Users/travis/.julia/packages/GAP/tkivN/src/GAP.jl:289
> ```
> 
> The other tests pass. What is going on here?

Does Julia 1.4.0 come before or after Julia 1.3? On my laptop I have Julia 1.4.0, and the tests work.

wbhart commented:
After. Julia 1.4.0 = 1.4

ThomasBreuer commented:
@GDeFranceschi Yes, I had also kicked off Julia 1.3 from my machine some time ago.
Now I have installed it again, and the situation is as follows:
Julia 1.3.1 can load both GAP.jl 0.4.2 and the dev version of GAP.jl without problem,
but Julia 1.3.1 runs into an error when it tries to load GAP.jl 0.4.3 (the latest released version).
Thus the problem is on the side of GAP.jl (see oscar-system/GAP.jl/issues/509).

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -137,7 +138,13 @@ end
+   # but GAP's `IsSubset` check is not as cheap as one wants;
+   # there is an `IsSubset` method that checks for identity,
+   # but it is not always the first choice.
+   if GAP.Globals.IsIdenticalObj(x.X, y.X)
```
```suggestion
   if x.X === y.X
```

ThomasBreuer replied:
Yes, this is a good idea.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -575,19 +582,35 @@ normaliser = normalizer
 """
 function pcore(G::GAPGroup, p::Int64)
-   if !GAP.Globals.IsPrime(p)
+   if !GAP.Globals.IsPrimeInt(p)
```
perhaps we should rather call the relevant function from Oscar/Nemo/Hecke/AbstractAlgebra

ThomasBreuer replied:
Yes. (I had suggested this in the description.)

fingolfin commented on src/Groups/homomorphisms.jl:
```diff
@@ -239,11 +239,13 @@ end
     cokernel(f::GAPGroupHomomorphism)
-Return the cokernel of `f`.
+Return the cokernel of `f`, that is, the quotient of the codomain of `f`
+by the normal closure of the image.
```
That's the definition of cokernel in category theory. It is however *not* the definition GAP uses for "cokernel" (which is the kernel of the "reverse" generalized map). 

I think it's good to have it this way, but I'd suggest we add a warning about this in the doc string, for long time GAP users. Of course that's beyond the scope of this PR, but it'd be nice nevertheless.

ThomasBreuer replied:
Judging from the code (modulo the error), this seems to be the intended definition.
But of course we have to decide what we want to model, and then change the code if necessary.

If we do eventually want to support a machinery of multi-valued mappings like in GAP then we may take a definition as in GAP (`CoKernelOfMultiplicativeGeneralMapping`), and then just delegate to the relevant GAP functions.
If only homomorphisms are of interest then we may take the proposed definition.

Perhaps the most elegant solution is to use more specific variable names, in order to allow for several definitions.

fingolfin replied:
The issue here is simply one of names: "cokernel" has a well-established meaning in category theory, and I am pretty sure most mathematicians will only know that meaning. While the GAP cokernel is, as far as I can tell, a completely unrelated object. So the question is, what to call the Oscar equivalent of `CoKernelOfMultiplicativeGeneralMapping`? Perhaps @mohamed-barakat has a good idea?

mohamed-barakat replied:
The category of groups has cokernels. That the cokernel object is isomorphic to the quotient of the range modulo the normal-closure of the set-theoretic image is not a definition but a proposition.

I am puzzled by the documentation of `CoKernelOfMultiplicativeGeneralMapping` maybe because I am not sure if I understand the meaning of a `MultiplicativeGeneralMapping`. What I believe to understand:
* a general mapping in `GAP` is simply a relation where the first factor of the product is called source and the second range
* a multiplicative general mapping is a relation which takes the extra structure into account, i.e., if G and H are groups then the relation must be a subgroup of G x H.

Right?

fingolfin replied:
Right. And the "Cokernel" (in the GAP sense) is simply the "mirror" of the kernel; i.e. the kernel of the generalized inverse map. Or put yet another way: the kernel is the subset of all elements (g,1) in the relation, while the cokernel is the subset  of elements (1,g) -- which for an "honest" homomorphism of course only consists of the single element (1,1)

mohamed-barakat replied:
General mappings form a category. However, this category has neither initial nor terminal objects and hence no zero objects and no notion of a zero morphism. This also means that there is no notion of kernel or cokernel. If you wish I can think of a categorical characterization of the set-theoretic subobjects which are called Kernel and CoKernel of multiplicative general mappings in GAP.

ThomasBreuer committed:
follow two suggestions by @fingolfin

- `===` instead of `GAP.Globals.IsIdenticalObj`
- `isprime` instead of `GAP.Globals.IsPrime`

fingolfin reviewed (approved):
Let's just merge this now.
